File: src_py/Chapter3/technology.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Technology模块 - Technology类的Python实现
转换自Java版本的Technology.java
"""

import math

"""
@author Gianluca Capone and Davide Sgobba
Python转换

此类包含所有标识两种半导体技术的变量和参数：晶体管和微处理器
"""
class Technology:
    
    def __init__(self, parameters):
        """
        构造函数
        
        Args:
            parameters: 参数数组
        """
        # 参数
        self.label = parameters[1]                    # 技术标签 - 直接使用字符串 'TR' 或 'MP'
        self.cheap_lim = float(parameters[2])         # 成本技术前沿 (LAMBDA-CH_k)
        self.perf_lim = float(parameters[3])          # 性能技术前沿 (LAMBDA-PE_k)
        self.diagonal = math.sqrt(                    # 对角线：从原点到前沿的距离 (d_k)
            self.cheap_lim**2 + self.perf_lim**2
        )
        self.min_init_bud = float(parameters[4])      # 企业初始预算的最小值 (B-0_f,k)
        self.range_init_bud = float(parameters[5])    # 企业初始预算的值范围 (B-0_f,k)
        self.num_of_firms = int(parameters[6])        # 潜在企业数量 (F_k) 

File: src/Chapter3/Technology.java
package Chapter3;

/*                                                                            *
 *  @author Gianluca Capone and Davide Sgobba                                 *
 *                                                                            */

/* This class contains all variables and parameters that identify the two     *
 * semiconductor technologies: transistors and microprocessors                */
class Technology {

    // PARAMETERS
    String label;
    /* Label of the technology                                                */
    double cheapLim;
    /* Cheapness technological frontier (LAMBDA-CH_k)                         */
    double perfLim;
    /* Performance technological frontier (LAMBDA-PE_k)                       */
    double diagonal;
    /* Diagonal: distance from the origin to the frontier (d_k)               */
    double minInitBud;
    /* Minimum value of initial firm budget (B-0_f,k)                         */
    double rangeInitBud;
    /* Range of value of initial firm budget (B-0_f,k)                        */
    int    numOfFirms;
    /* Number of potential firms (F_k)                                        */
    
    // CONSTRUCTOR
    Technology(String[] PARAMETERS ) {
        label       = PARAMETERS[1];
        cheapLim    = Double.parseDouble(PARAMETERS[2]);
        perfLim     = Double.parseDouble(PARAMETERS[3]);
        diagonal    = Math.sqrt(cheapLim * cheapLim + perfLim * perfLim);
        minInitBud  = Double.parseDouble(PARAMETERS[4]);
        rangeInitBud = Double.parseDouble(PARAMETERS[5]);
        numOfFirms  = Integer.parseInt(PARAMETERS[6]);
    }
    
}